xun/functions/util.py

```diff
@@ -275,19 +275,70 @@ def shape_to_ast_tuple(shape):
 
 def stmt_introduced_names(stmt):
     """
-    Return a list of all names introduced by executing the statement
+    Return a list of all names introduced by executing the statement,
+    assuming the statement is run in isolation
     """
+    def add_names_from_body(name_list, body):
+        for body_stmt in body:
+            for item in stmt_introduced_names(body_stmt):
+                name_list.add(item)
+
+    if isinstance(stmt, (ast.FunctionDef, ast.AsyncFunctionDef)):
+        return frozenset({stmt.name})
+    if isinstance(stmt, ast.ClassDef):
+        return frozenset({stmt.name})
+    if isinstance(stmt, ast.Return):
+        return frozenset()
+    if isinstance(stmt, ast.Delete):
+        return frozenset()
+    if isinstance(stmt, ast.Assign):
+        targets = stmt.targets
+        return assignment_target_introduced_names(targets)
+    if isinstance(stmt, ast.AugAssign):
+        return frozenset()
+    if isinstance(stmt, ast.AnnAssign):
+        return frozenset({stmt.target.id})
+    if isinstance(stmt, (ast.For, ast.AsyncFor)):
+        names = {stmt.target.id}
+        add_names_from_body(names, stmt.body)
+        return frozenset(names)
```

Comment:
We need to consider the case where the `stmt.target` is a tuple, e.g. `for i,j in d.items():`.

```suggestion
    def names_in_target(target):
        names = set()
        for elt in target.elts:
            if isinstance(elt, ast.Name):
                names.add(elt.id)
            elif isinstance(elt. ast.Tuple):
                names.update(names_in_target(elt)
            else:
                raise TypeError(f'Invalid AST type {type(elt)}')
        return names
        
    if isinstance(stmt, (ast.For, ast.AsyncFor)):
        names = names_in_target(stmt.target)
        add_names_from_body(names, stmt.body)
        return frozenset(names)
```